doc/neps/nep-0052-python-api-cleanup.rst

```diff
@@ -0,0 +1,331 @@
+  and other issues for ``maximum_sctype`` and related functions).
+- Business day functionality can likely be removed (unclear if it needs
+  splitting out like was done for ``np.financial``).
+- The ``np.compat`` namespace, used during the Python 2 to 3 transition, will be removed.
+- Functions that are narrow in scope, with very few public use-cases,
+  will be removed.  See, e.g.
+  ``real_if_close`` (`gh-11375 <https://github.com/numpy/numpy/issues/11375>`__).
+  These will have to be identified manually and by issue triage.
+
+New namespaces are introduced for warnings/exceptions (``np.exceptions``) and
+for dtype-related functionality (``np.types``). NumPy 2.0 is a good opportunity
+to populate these submodules from the main namespace.
+
+Functionality that is widely used but has a preferred alternative may either be
+deprecated (with the deprecation message pointing out what to use instead) or
+be hidden by not including it in ``__dir__``. In case of hiding, a ``..
+legacy::`` directory may be used to mark such functionality in the
+documentation.
+
+A test will be added to ensure limited future growth of all namespaces; i.e.,
+every new entry will need to be explicitly added to an allow-list.
+
+
+Cleaning up the submodule structure
+```````````````````````````````````
+
+We will clean up the NumPy submodule structure, so it is easier to navigate.
+When this was discussed before (see
+`MAINT: Hide internals of np.lib to only show submodules <https://github.com/numpy/numpy/pull/18447>`__)
+there was already rough consensus on that - however it was hard to pull off in
+a minor release.
+
+A basic principle we will adhere to is "one function, one location". Functions
+that are exposed in more than one namespace (e.g., many functions are present
+in ``numpy`` and ``numpy.lib``) need to find a single home.
+
+We will reorganize the API reference guide along main and submodule namespaces,
+and only within the main namespace use the current subdivision along
+functionality groupings. Also by "mainstream" and special-purpose namespaces.
+Details TBD, something like:
+
+::
+
+    # Regular/recommended user-facing namespaces for general use. Present these
+    # as the primary set of namespaces to the users.
+    numpy
+    numpy.exceptions
+    numpy.fft
+    numpy.linalg
+    numpy.ma
+    numpy.polynomial
+    numpy.random
+    numpy.testing
+    numpy.typing
+
+    # Special-purpose namespaces. Keep these, but document them in a separate
+    # grouping in the reference guide and explain their purpose.
+    numpy.array_api
+    numpy.ctypeslib
+    numpy.emath
+    numpy.f2py  # only a couple of public functions, like `compile` and `get_include`
+    numpy.math
+    numpy.lib.stride_tricks
+    numpy.rec
+    numpy.types
+
+    # Legacy (prefer not to use). Third grouping in the reference guide.
+    numpy.char
+    numpy.distutils
+    numpy.matrixlib  # or deprecate?
+
+    # To remove
+    numpy.compat
+    numpy.core  # rename to _core
```

Comment:
`core` might warrant a bit more discussion. Certainly, the build-related stuff should be hidden, but things like `arrayprint` and `einsumfunc` probably not. Where would those go? Maybe a good general principle could be that if one does `np.<function>?` in `ipython`, it gives where the function is defined, and I think it makes most sense if that module is not private (even if it is not meant to be imported directly; it should indicate what is public with `__all__`). If considered too numpy-specific, something like `np.utils` might be fine.

But just my 2¢; I'm not sure if this would be consistent with what has been done in scipy, and consistency with that is probably more important!

Reply:
There's very little there that's not in the main namespace:

```python
>>> objs_main = set([s for s in dir(np) if not s.startswith('_')])
>>> objs_core = set([s for s in dir(np.core) if not s.startswith('_')])
>>> objs_core - objs_main
{'arrayprint', 'defchararray', 'shape_base', 'function_base', 'numerictypes', 'numeric', 'getlimits', 'warnings', 'unicode', 'MachAr', 'einsumfunc', 'umath', 'fromnumeric', 'multiarray', 'overrides', 'records', 'long'}
```

`arrayprint` is a module, and I'm not sure there's anything in there that needs to be exposed elsewhere. We have `np.array_str`, `np.set_printoptions` & co already.

Same for `einsumfunc`, it's a module, and `np.einsum` and np.einsum_path` are already in the main namespace.


> Maybe a good general principle could be that if one does `np.<function>?` in `ipython`, it gives where the function is defined, and I think it makes most sense if that module is not private

I don't agree with this part. We set `__module__` to the public place with `@set_module('numpy')`, and it's perfectly fine for IPython to show a private file as the source location with `?`. There is no reason for original source code to only be in public submodules, that is unworkable.

Reply:
For `arrayprint`, I think the classes defined there can be quite useful, so that suggest its own module, but perhaps it is actually best not to expose the API too much.

On the naming, I'm just being biased by the astropy usage, where one always imports from the top of a submodule, and things below are considered implementation. I find directories full of `_*.py` files somewhat strange, but it definitely makes sense to just stick to the new scipy scheme; consistency is important.

Reply:
> For `arrayprint`, I think the classes defined there can be quite useful, so that suggest its own module, but perhaps it is actually best not to expose the API too much.

That's all not in `__all__` and was clearly designed to be private. I think it's best to have users who want to use that to know that that's the case - they can still import from `_arrayprint` (with underscore) but then it's clear what they're doing. And they may decide they're better off vendoring that code instead, if they do indeed need it.

And yes indeed, I'd like to avoid exposing it as new public API.

Reply:
> where one always imports from the top of a submodule, and things below are considered implementation.

I don't think I understand what that means. What is "top" vs. not-top? Does it have to do with `__all__`? Or otherwise, how does an AstroPy user know what is public vs. private?

Reply:
Astropy is a bit more like scipy, in that almost everything is one level down (`astropy.units`, `astropy.time`, etc.), and it is set up such that one imports from those, never further down. But so far no explicit naming convention and basically no files starting with underscores (though this may change, we're discussing how to separate public vs private API as well; but it could be as simple as having `__all__` in the corresponding `__init__` to make things a bit clearer; overall, I don't feel it has been much of a problem).